modules/packages/Sort.chpl

```diff
@@ -1319,12 +1570,1106 @@ module RadixSortHelp {
+    iter classify(A, start_n, end_n, criterion, startbit) {
+      var cur = start_n;
+      while cur <= end_n-(classifyUnrollFactor-1) {
+        for /*param*/ j in 0..classifyUnrollFactor-1 {
+          yield (cur+j, bucketForRecord(A[cur+j], criterion, startbit));
+        }
+        cur += classifyUnrollFactor;
+      }
+      while cur <= end_n {
+        yield (cur, bucketForRecord(A[cur], criterion, startbit));
+        cur += 1;
+      }
+    }
+  }
+}
+
+pragma "no doc"
+module ShallowCopy {
+
+  // These shallow copy functions "move" a record around
+  // (i.e. they neither swap nor call a copy initializer).
+
+  // TODO: move these out of the Sort module and/or consider
+  // language support for it.
+
+  // TODO: These shallowCopy functions should handle Block,Cyclic arrays
+  inline proc shallowCopy(ref A, dst, src, nElts) {
+    // Ideally this would just be
+    //A[dst..#nElts] = A[src..#nElts];
+
+    if boundsChecking {
+      assert(nElts > 0);
+      assert(A.domain.contains(dst..#nElts));
+      assert(A.domain.contains(src..#nElts));
+    }
+
+    if A._instance.isDefaultRectangular() {
+      var size = (nElts:size_t)*c_sizeof(A.eltType);
+      c_memcpy(c_ptrTo(A[dst]), c_ptrTo(A[src]), size);
+    } else {
+      var ok = chpl__bulkTransferArray(/*dst*/ A, {dst..#nElts},
+                                       /*src*/ A, {src..#nElts});
+      if !ok {
+        halt("bulk transfer failed in sorting");
+        for i in vectorizeOnly(0..#nElts) {
+          __primitive("=", A[dst+i], A[src+i]);
+        }
+      }
+    }
+  }
+  inline proc shallowCopy(ref DstA, dst, ref SrcA, src, nElts) {
+    // Ideally this would just be
+    //DstA[dst..#nElts] = SrcA[src..#nElts];
+
+    if boundsChecking {
+      assert(nElts > 0);
+      assert(DstA.domain.contains(dst..#nElts));
+      assert(SrcA.domain.contains(src..#nElts));
+    }
+
+    if DstA._instance.isDefaultRectangular() &&
+       SrcA._instance.isDefaultRectangular() {
+      var size = (nElts:size_t)*c_sizeof(DstA.eltType);
+      c_memcpy(c_ptrTo(DstA[dst]), c_ptrTo(SrcA[src]), size);
+    } else {
+      var ok = chpl__bulkTransferArray(/*dst*/ DstA, {dst..#nElts},
+                                       /*src*/ SrcA, {src..#nElts});
+      if !ok {
+        halt("bulk transfer failed in sorting");
+        for i in vectorizeOnly(0..#nElts) {
+          __primitive("=", DstA[dst+i], SrcA[src+i]);
+        }
+      }
+    }
+  }
+  inline proc parallelShallowCopy(ref DstA, dst, ref SrcA, src, nElts) {
+    const nTasks = if dataParTasksPerLocale > 0
+                   then dataParTasksPerLocale
+                   else here.maxTaskPar;
+
+    halt("buggy");
+    // TODO: something about this code is wrong right now.
+    const blockSize = divceil(nElts, nTasks);
+    const nBlocks = divceil(nElts, blockSize);
+    coforall tid in 0..#nTasks {
+      var start = tid * blockSize;
+      var n = blockSize;
+      if start + n > nElts then
+        n = nElts - start;
```

Comment:
If `nTasks > nElts + 1`, you'll end up with a negative `n` here  -- is that what's wrong?